### DISCUSSION TITLE
Compiling questions about using fluid properties module

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#22007] [Q&A General] [answered]
Hello, everyone.
I am trying to use MOOSE to simulate two-phase flow based on the basic homogeneous equations, and in the process, I need to calculate void fraction, densities and the saturate temperature and stuffs like that.
Then I thinke it's a good choice to develop a outside module to calculate the properties, like, creating a C++ file with functions, and  transfer the input variables in then get parameters I want.
And then I decide to look into the FLUID PROPERTIES module (Water97FluidProperties, to be specific), to import the desired properties parameters into the Material Object, identicial to the process presented in the page below.
[https://mooseframework.inl.gov/modules/fluid_properties/index.html](url)
Then I turn the module selection (FP only) in the Makefile to "yes". But it says that the Material file I want to use is not registered and the process goes wrong.
Then I turn all the modules on in the Makefile, and everything goes well.
So, does that mean the FP modules can only be used with some other modules imported or something? If I just want to use it as a calculation module, is there anything more I should take care?
Could you help me with this problem, thanks a lot.

**Comment by GiudGiud** [comment, marked as answer]
Hello

you should be able to import just the fluid properties module. Maybe something isn’t set up quite right. 

 Could you please try building your application with the Navier Stokes module? That should include the fluid properties module

Guillaume

**Reply by Raven-pro** [reply]
Thank you! Sorry for not being able to reply in time.
I have revised my setting and found what is wrong(mistyping of my code mostly), after the debugging the code could run successfully.
But I ran into another problem, the parameters supposed to be calculated in the Material module cannot transfer to the kernels(I tried to use `std::cout << `to see if  I get parameters wrong, and they were wrong indeed.), causing the disconvergence of the simulation.
Below is my input file. Could you help me find out what's wrong? Thanks a lot!

```C++
[Mesh]
    type = GeneratedMesh
    dim = 1
    nx = 10
    xmax = 1.0
    xmin = 0
[]

[Modules]
    [FluidProperties]
        [iapws]
            type = Water97FluidProperties
        []
    []
[]

[Variables]
    [P]
        order = FIRST
        family = LAGRANGE
        initial_condition = 15500000.0
    []
    [W]
        order = FIRST
        family = LAGRANGE
        initial_condition = 0.3911
    []
    [H]
        order = FIRST
        family = LAGRANGE
        initial_condition = 1133820.0
    []
[]

[Materials]
    [fp_mat]
        type = IAPWS97Material
        pm = P
        hm = H
        fp = iapws
    []
[]

[Kernels]
    [ContinumPressureTime]
        type = ContinumPressureTime
        variable = P
        hm = H
    []
    [ContinumMassGrad]
        type = ContinumMassGrad
        variable = P
        grad_w = W
        velocity = "1.0 1.0 1.0"
    []
    [MomentumMassTime]
        type = MomentumMassTime
        variable = W
    []
    [MomentumMassGrad]
        type = MomentumMassGradCoupled
        variable = W
        p = P
        hm = H
        velocity = "1.0 1.0 1.0"
    []
    [MomentumPressureGrad]
        type = MomentumPressureGradCoupled
        variable = W
        p = P
        hm = H
        grad_p = P
        velocity = "1.0 1.0 1.0"
    []
    [MomentumMassSource]
        type = MomentumMassSource
        variable = W
        p = P
        hm = H
    []
    [MomentumFrictionSource]
        type = MomentumFrictionSource
        variable = W
        p = P
        hm = H
    []
    [EnergyEnthalpyTime]
        type = EnergyEnthalpyTime
        variable = H
        p = P
    []
    [EnergyEnthalpyGrad]
        type = EnergyEnthalpyGrad
        variable = H
        w = W
        velocity = "1.0 1.0 1.0"
    []
    [EnergyHeatSource]
        type = EnergyHeatSource
        variable = H
    []
    [EnergyPressureTime]
        type = EnergyPressureTime
        variable = H
        v = P
    []
    [EnergyFrictionSource]
        type = EnergyFrictionSource
        variable = H
        w = W
        p = P
    []
[]

[BCs]
    [bc1]
        type = DirichletBC
        variable = H
        value = 1133820.0
        boundary = "left"
    []
    [bc2]
        type = DirichletBC
        variable = P
        value = 15500000.0
        boundary = "right"
    []
    [bc3]
        type = DirichletBC
        variable = W
        value = 0.3911
        boundary = "left"
    []
[]

#[VectorPostprocessors]
  #[H]
    #type = NodalValueSampler
    #variable = H
    #boundary = 'bottom'
    #block = 0
    #execute_on = 'initial timestep_end' 
    #use_displaced_mesh = true
    #sort_by = x
  #[]
  #[P]
    #type = NodalValueSampler
    #variable = P
    #boundary = 'bottom'
    #block = 0
    #execute_on = 'initial timestep_end' 
    #use_displaced_mesh = true
    #sort_by = x
  #[]
#[]

[Problem]
    type = FEProblem
    coord_type = XYZ
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
    type = Transient
    solve_type =  PJFNK
    automatic_scaling = false
    scheme = bdf2
    petsc_options_iname = '-pc_type -pc_hypre_type'
    petsc_options_value = 'hypre boomeramg'
    num_steps = 10000
    #start_time = 0.0
    dt = 0.00001
    #end_time = 1.0
    steady_state_tolerance = 0.1
    steady_state_detection = true
[]

[Outputs]
    exodus = true
    csv = true
[]

```

**Reply by Raven-pro** [reply]
Another thing I want to ask about is, I noticed that all the userobjects files of the module FLUID PROPERTIES are registered to FluidPropertiesApp.
Does that mean, when I am compiling, there will be built both my application and the properties application, and I can use the parameters calculated in the FluidPropertiesApp to perform the simulation in my app, some kind of coupling?

**Reply by GiudGiud** [reply]
Hello

Objects registered to the fluid properties app will be available in any app that includes the FLUID_PROPERTIES module. You can look at the Makefile to see which modules are included. 

What is this IAPWS97 material? It's not part of MOOSE. 
Why are you not using the materials in MOOSE (such as FluidPropertiesMaterialPT, or GeneralFluidProps) to compute the fluid properties from the iapws object?

Guillaume

**Reply by Raven-pro** [reply]
Yes, `IAPWS97Material` is the file I created myself, and I have also added some functions in the `Water97FluidProperties` file. The material file is in `moose/modules/fluid_properties/src/materials/`. 
The codes are as follows. 
```C++
#include "IAPWS97Material.h"
#include "Water97FluidProperties.h"

registerMooseObject("FluidPropertiesApp", IAPWS97Material);

InputParameters
IAPWS97Material::validParams()
{
    InputParameters params = Material::validParams();
    params.addRequiredCoupledVar("pm", "Pressure mixture");
    params.addRequiredCoupledVar("hm", "Enthalpy mixture");
    params.addRequiredParam<UserObjectName>("fp", "The name of the user object for fluid properties");
    params.addClassDescription("Compute fluid properties using formulations");
    return params;
}

IAPWS97Material::IAPWS97Material(const InputParameters & parameters): 
Material(parameters),
_Enth(coupledValue("hm")),
_Pressure(coupledValue("pm")),
/// Properties to be calculated
_hg(declareProperty<Real>("hg")),
_hl(declareProperty<Real>("hl")),
_vg(declareProperty<Real>("vg")),
_vl(declareProperty<Real>("vl")),
_T(declareProperty<Real>("temperature")),
_alpha(declareProperty<Real>("alpha")),
_kg(declareProperty<Real>("kg")),
_kl(declareProperty<Real>("kl")),
_rhog(declareProperty<Real>("rhog")),
_rhol(declareProperty<Real>("rhol")),
_rhom(declareProperty<Real>("rhom")),

_fp(getUserObject<Water97FluidProperties>("fp"))
{}

IAPWS97Material::~IAPWS97Material(){}

void
IAPWS97Material::computeQpProperties()
{
    _T[_qp] = _fp._Tsat_P(_Pressure[_qp]);
    _hg[_qp] = _fp._hg_sat_P(_Pressure[_qp]);
    _hl[_qp] = _fp._hl_sat_P(_Pressure[_qp]);
    _vg[_qp] = _fp._vg_sat_P(_Pressure[_qp]);
    _vl[_qp] = _fp._vl_sat_P(_Pressure[_qp]);
    _rhog[_qp] = 1/_vg[_qp];
    _rhol[_qp] = 1/_vl[_qp];
    _alpha[_qp] = _fp._hmm_alpha(_Pressure[_qp], _hg[_qp], _hl[_qp], _rhog[_qp], _rhol[_qp]);
    _rhom[_qp] = _alpha[_qp] * _rhog[_qp] + (1 - _alpha[_qp]) * _rhol[_qp];
    Real epsilon = 100;
    _kg[_qp] = ((1/_fp._vg_sa(t_P(_Pressure[_qp] + epsilon)) - (1/_fp._vg_sat_P(_Pressure[_qp])))/epsilon;
    _kl[_qp] = ((1/_fp._vl_sat_P(_Pressure[_qp] + epsilon)) - (1/_fp._vl_sat_P(_Pressure[_qp])))/epsilon;
}
```

**Reply by Raven-pro** [reply]
But the parameters supposed to be transfered to the main App are wrong and constant(like _T[_qp], is always value 25.491, and _kg[_qp] is always 1.15236e-241). I tried changing the expression of the parameters(like setting value _T[_qp] = 1) and it doesn't work.

**Reply by GiudGiud** [reply]
are you storing the material properties as references?
if not you are making local copies, which is why it doesnt work

**Reply by Raven-pro** [reply]
I recompiled the App and then it worked. Maybe I should use `make clean` each time I compile the App. 
Thanks a lot for your patient replies!
